sphinx/util/inspect.py

```diff
@@ -87,7 +87,12 @@ def getall(obj: Any) -> Sequence[str] | None:
 
 def getannotations(obj: Any) -> Mapping[str, Any]:
     """Get __annotations__ from given *obj* safely."""
-    __annotations__ = safe_getattr(obj, '__annotations__', None)
+    if (sys.version_info >= (3, 10, 0)) or (not isinstance(obj, type)):
+        __annotations__ = safe_getattr(obj, '__annotations__', None)
+    else:
+        # Workaround for bugfix not available until python 3.10 as recommended by docs
+        # https://docs.python.org/3.10/howto/annotations.html#accessing-the-annotations-dict-of-an-object-in-python-3-9-and-older
+        __annotations__ = obj.__dict__.get('__annotations__', None)
```

Comment:
Not all objects have a `__dict__` (some ill-formed types may not necessarily have one) so I'd suggest yousing `getattr(obj, '__dict__', {})` for this one (or use `safe_getattr` which *should* be fine).

Reply:
Will look more at this -- I did use the exact code recommended by the python contributor team though

Reply:
I want to be careful because the whole point of this PR is a corner case where safe_getattr definitely does not work, and I'm not sure how to test a malformed type?

Reply:
The official doc also mentions 

> Note that some exotic or malformed type objects may not have a __dict__ attribute, so for extra safety you may also wish to use [getattr()](https://docs.python.org/3.10/library/functions.html#getattr) to access __dict__.

Now for actual testing, we can just ignore it. I am not aware of any construction in plain Python that makes the `__dict__` of a class disappear, even with metaclass hacks. It usually happens when a class is written in C (mypyc native classes may not have a `__dict__` iirc, but I'm not sure that this portion of the code could be triggered with them).

Anyway, if you want to be on the safe side, using `getattr(obj, '__dict__', {})` is sufficient.

Reply:
So, use getattr to get the dict, but not add a test case for the corner case of the corner case?